mne/export/_edf.py

```diff
@@ -0,0 +1,212 @@
+# -*- coding: utf-8 -*-
+# Authors: MNE Developers
+#
+# License: BSD-3-Clause
+
+import numpy as np
+
+from ..utils import _check_edflib_installed, warn
+_check_edflib_installed()
+from EDFlib.edfwriter import EDFwriter  # noqa: E402
+
+
+def _try_to_set_value(header, key, value, channel_index=None):
+    """Helper function to set key/value pairs in EDF/BDF header."""
+    # all EDFLib set functions are set<X>
+    # for example "setPatientName()"
+    func_name = f'set{key}'
+    func = getattr(header, func_name)
+
+    # some setter functions are indexed by channels
+    if channel_index is None:
+        return_val = func(value)
+    else:
+        return_val = func(channel_index, value)
+
+    # a return value of 0 indicates an error
+    if return_val != 0:
+        raise RuntimeError(f"Setting {key} with {value} "
+                           f"returned an error")
+
+
+def _export_raw(fname, raw, physical_range, fmt):
+    """Export Raw objects to EDF or BDF files.
+
+    TODO:
+    - if Info stores transducer information, allow writing here.
+    - if Info stores techniciain information, allow writing here.
+    """
+    phys_dims = 'uV'
+
+    if fmt == 'bdf':
+        digital_min = -8388607
+        digital_max = 8388607
+    elif fmt == 'edf':
+        digital_min = -32767
+        digital_max = 32767
+
+    # load data first
+    raw.load_data()
+
+    # remove extra epoc and STI channels
+    drop_chs = ['epoc']
+    if not (raw.filenames[0].endswith('.fif')):
+        drop_chs.append('STI 014')
+
+    ch_names = [ch for ch in raw.ch_names if ch not in drop_chs]
+    n_chs = len(ch_names)
+    n_times = raw.n_times
+    sfreq = int(raw.info['sfreq'])
+    n_secs = n_times / sfreq
+
+    # get any filter information applied to the data
+    lowpass = raw.info['lowpass']
+    highpass = raw.info['highpass']
+    linefreq = raw.info['line_freq']
+    filter_str_info = f"HP:{highpass}Hz LP:{lowpass}Hz N:{linefreq}Hz"
+
+    # get data in uV
+    # XXX: what about non-EEG data?... How to handle?
+    units = dict()
+    units['eeg'] = 'uV'
+    units['ecog'] = 'uV'
+    units['seeg'] = 'uV'
+
+    # get the entire dataset in uV
```

Comment:
them it makes no sense to store anything else than EEG, sEEG or ECoG data in edf. I think that EDF file actually have a trigger channel (like a stim channel in MEG)